tools/erc-repository-indexer/docs/registry_design.md

```diff
@@ -0,0 +1,149 @@
+  - `MIRROR_NODE_URL`: API URL for the Hedera mirror node.
+  - `STARTING_POINT`: a starting contract ID or a potential `next` pointer included in mirror node REST response (TBD).
+
+### Implementation Workflow
+
+1. **Initialization**:
+
+   - Load environment variables and configuration parameters.
+
+2. **Identifying ERCs**:
+
+   a. **Retrieve Contract Batches**
+
+   - Use the Hedera Mirror Node API to index the entire network, retrieving contracts in ascending order (from oldest to newest):
+
+     - `https://{{network}}.mirrornode.hedera.com/api/v1/contracts?order=asc`
+
+   b. **Fetch Contract Details and Parse Bytecode**
+
+   - For each contract:
+     - Retrieve contract details:
+       - `https://{{network}}.mirrornode.hedera.com/api/v1/contracts/{contractIdOrEvmAddress}`
+     - Extract and parse the bytecode from the contract detail object.
+       _Note: Utilize concurrent requests to optimize execution efficiency._
+
+   c. **Perform Interface Signature Matching**
+
+   - Analyze the bytecode for compliance with ERC-20 or ERC-721 standards using the [Contract.isErc()](https://github.com/acuarica/evm/blob/402028ca8c3a33dbb8498f0200d9af2efbf4f792/src/index.ts#L153-L158) method from the [SEVM library](https://www.npmjs.com/package/sevm). This method analyzes the contract bytecode and matches function signatures to confirm compliance with the ERC-20 and ERC-721 standards.
+
+   - Identify the following function selectors in the bytecode:
+
+     - **ERC-20 Selectors**:
+
+       - `0x18160ddd`: `totalSupply()`
+       - `0x70a08231`: `balanceOf(address)`
+       - `0xa9059cbb`: `transfer(address,uint256)`
+       - `0x23b872dd`: `transferFrom(address,address,uint256)`
+       - `0x095ea7b3`: `approve(address,uint256)`
+       - `0xdd62ed3e`: `allowance(address,address)`
+
+     - **ERC-721 Selectors**:
+       - `0x6352211e`: `ownerOf(uint256)`
+       - `0x42842e0e`: `safeTransferFrom(address,address,uint256)`
+       - `0xb88d4fde`: `safeTransferFrom(address,address,uint256,bytes)`
+       - `0x095ea7b3`: `approve(address,uint256)`
+       - `0x081812fc`: `getApproved(uint256)`
+       - `0xa22cb465`: `setApprovalForAll(address,bool)`
+       - `0x70a08231`: `balanceOf(address)`
+       - `0xe985e9c5`: `isApprovedForAll(address,address)`
+       - `0x01ffc9a7`: `supportsInterface(bytes4)`
+       - `0x23b872dd`: `transferFrom(address,address,uint256)`
+
+d. **Handle Pagination**
+
+- Since the `/contracts` endpoint has a limit of 100 records per request, use the `next` pointer provided in the response to recursively fetch subsequent batches and comprehensively index the network.
+- Write the `next` pointer to disk for persistence, allowing the indexing process to resume from the last recorded point if interrupted or to accommodate future runs when additional contracts are deployed on the network.
+
+3. **Validation (Optional)**:
+
+   - Perform `eth_call` on detected contracts to confirm:
```

Comment:
We could bypass the relay overload by executing this call straight on the mirror node web3 module. The doc is [here](https://mainnet-public.mirrornode.hedera.com/api/v1/docs/#/contracts/contractCall).